src/Mod/Draft/draftfunctions/move.py

```diff
@@ -215,4 +218,208 @@ def copy_moved_edge(object, edge_index, vector):
+            outer_wire = w
+    
+    new_wires = [outer_wire]
+    for w in wires:
+        if w != outer_wire:
+            new_wires.append(w)
+    
+    return new_wires
+
+    
+def parse_sub_shapes(shape, selected_subelements, vector):
+    """ Parse the subshapes of the given shape in order to
+    find modified shapes and substitute them to the originals.
+    """
+    sub_shapes = []
+    touched_subshapes = []
+    if shape.SubShapes:
+        for sub_shape in shape.SubShapes:
+            new_sub_shape, touched_subshape = parse_shape(sub_shape, selected_subelements, vector)
+            sub_shapes.append(new_sub_shape)
+
+            if touched_subshape:
+                touched_subshapes.append(2)
+            else:
+                touched_subshapes.append(0)
+
+    if 0 in touched_subshapes and 2 in touched_subshapes:
+        # only some subshapes touched
+        touched = 1
+    elif 2 in touched_subshapes:
+        # all subshapes touched
+        touched = 2
+    elif 0 in touched_subshapes:
+        # no subshapes touched
+        touched = 0
+
+    return sub_shapes, touched
+
+
+def moveSubElements(obj, sub_objects_names, vector):
+    """moveSubElements(obj, sub_objects_names, vector)
+    
+    Move the given object sub_objects according to a vector or crates an new one
+    if the object is not a Part::Feature.
+
+    Parameters
+    ----------
+    obj : the given object
+
+    sub_objects_names : list of subelement names
+        A list of subelement names to identify the subelements to move.
+
+    vector : Base.Vector
+        Delta Vector to move subobjects from the original position. 
+
+    Return
+    ----------
+    shape : Part.Shape
+        Return the new Shape or None.
+    """
+
+    import Part
+
+    if not isinstance(sub_objects_names, list):
+        sub_objects_names = [sub_objects_names]
+
+    shape = obj.Shape
+    new_shape = None
+    if not shape.isValid():
+        return
+
+    selected_subelements = []
+    for sub_objects_name in sub_objects_names:
+        sub_object = obj.Shape.getElement(sub_objects_name)
+        selected_subelements.append(sub_object)
+        selected_subelements.extend(sub_object.Faces)
+        selected_subelements.extend(sub_object.Edges)
+        selected_subelements.extend(sub_object.Vertexes)
+
+    new_shape, touched = parse_shape(shape, selected_subelements, vector)
+    
+    if not new_shape.isValid():
+        should_fix = move_subelements_msgbox()
+        if should_fix:
+            new_shape.fix(0.001,0.001,0.001)
+
+    if new_shape:
+        if hasattr(obj, 'TypeId') and obj.TypeId == 'Part::Feature':
+            obj.Shape = new_shape
+        else:
+            new_obj = App.ActiveDocument.addObject("Part::Feature", "Feature")
+            new_obj.Shape = new_shape
+        return new_shape
+
+def move_subelements_msgbox():
+    if not App.GuiUp:
+        return False
+    msgBox = QtGui.QMessageBox()
```

Comment:
Personally, I think you should keep these functions inside `draftfunctions` entirely non-GUI. Any sort of GUI interaction should be programmed in the corresponding Gui Command, that is, `draftguitools/gui_move`. Otherwise you will be mixing non-GUI and GUI together, and it's harder to keep a clean programming interface.

In this case, the main `move_sub_elements` function could have a fourth parameter to control this.
```py
def move_sub_elements(obj, sub_objects_names, vector, fix_invalid=False)
```

Then
```py
    if not new_shape.isValid():
        should_fix = fix_invalid  # move_subelements_msgbox()
        if should_fix:
```

Reply:
yes, this looks better in the code, but not so smooth in the workflow... dont' you think?
Anyway no problem to change it.

Reply:
> but not so smooth in the workflow... dont' you think?

What is not smooth about it?

Reply:
Well, imagine you didnt' check the option to force fix: the fist result is an invalid shape, than you have to fix it manually or undo the operation and redo it with the option checked.
As it's implemented now, a message is prompt in case the result is an invalid shape, and the user is asked if he want it to be fixed. I think it's more user friendly.

Reply:
Correct. But then it is not a good "programming interface". A programming interface is something that you can script and run unattended. Adding pop-ups or dialogs will require manual intervention, the user clicking OK or Cancel; this means that this piece of code cannot be used properly as a scripting interface. In my opinion, this function does not need that; the graphical processing should occur elsewhere, for example, in the GUI command.

So, when you develop a new function you have to think about whether you want it to work in an unattended manner (like a proper programming interface), or whether it will be mostly used with user input.

For example, what happens if you want to automate moving many shapes? You suddenly won't be able to automate this because inside a loop, the function may be asking you questions every time.

The proper way of handling this would be to raise an exception or something else from the base function, and then trap that exception in the parent GUI code, which then would respond accordingly (show the text box or do something else).

Personally, I think the Draft code now in the new structure is quite clean because you can separate the behavior into GUI and non-GUI (programming interface); however, if you start adding dialogs like these, it will become tangled, and won't be as flexible.

Reply:
> The proper way of handling this would be to raise an exception or something else from the base function, and then trap that exception in the parent GUI code, which then would respond accordingly (show the text box or do something else).

ok! let's do it like this.